### PR TITLE
Allow Twig callable arguments to use camel or snake names

### LINKED ISSUE #3475: Named Arguments of Filters and Functions are not passed as-is and forced to snake_case
## Problem Description
I stumbled upon this super odd behaviour when implementing a custom function in Twig and realised that given the template
```twig
{{ foo(barValue='bar', booValue='boo') }}
```
the two named arguments, `barValue` and `booValue` will be passed to the `foo` function as `bar_value` and `boo_value`. This is due to the "name normalization" done in the [`CallExpression` on line 242](https://github.com/twigphp/Twig/blob/4a47f0f5b4bf2400ed1e3abecb486655e3970cdb/src/Node/Expression/CallExpression.php#L242) in the current 3.x branch. This has been introduced in a59dcde3, around 8 years ago and was part of the initial named arguments implementation.

As this behaviour is undocumented, technically unnecessary (named arguments are passed as array keys and PHP allows all strings to be array keys) and defnitely unexpected, I'm suggesting to classifiy this as a bug.

## Proposed Fix
To fix it without introducing backwards incompatible changes, I'm proposing a new option for functions and filters that is called `legacy_named_arguments` and that defaults to `true` in 3.x and can be set to `false` in upcoming major releases.

- Setting `legacy_named_arguments` to `true` means: keep the current behaviour. In my opinion it makes sense to trigger a `E_USER_DEPRECATED` when a variable key has been "normalized" (i. e. input key !== output key). However, this might have a very big impact on a lot of existing Symfony installations.

- Setting `legacy_named_arguments` to `false` means: leave the input argument as-is. The lexer decides whether a variable name is valid or not (which it already does implicitly [here](https://github.com/twigphp/Twig/blob/3.x/src/Lexer.php#L363)).

I am happy to contribute in a PR once you've confirmed the issue.

### PR DESCRIPTION
**Pull request by fabpot**
Closes #3475

Twig callables (functions/filters/tests) now accept snake and camel argument names independently of the name of the underlying PHP callable parameter names.

### CODE REVIEW

**Comment by smnandre** [comment]
I love how this implementation _dances_ around most of the differences between the  previous `normalize($str)` and the String  implementation.

I encountered many problems when the `snake` implementation changed in String...  so i tried to double-check many cases here :) 

> [!NOTE]
> i only looked at the code of the changed class, i am **not 100% sure** this concerns _real_ use cases (especially after the recent naming convention updates)
> 

## Results

A  ❌  in the `Same` column indicate a change of behaviour **before / after** this PR (but only some of them are "regressions" .... and again not sure about..

| Node Attribute | Callable Parameter | Same | norm(attr)   | norm(call)   | Valid Before? | node attr    | res = param  | camel(param) | snake(camel(param)) | Valid After? | 
| - | - | - | - | - | - | - | - | - | - | - |
 no_svg         |  noSVG          |  ✅  |  no_svg         |  no_svg         |  ✅  |  no_svg         |  noSVG          |  noSVG          |  no_svg         |  ✅  | 
 noSVG          |  no_svg         |  ❌  |  no_svg         |  no_svg         |  ✅  |  noSVG          |  no_svg         |  noSvg          |  no_svg         |  ❌  | 
 error_404      |  error404       |  ✅  |  error_404      |  error404       |  ❌  |  error_404      |  error404       |  error404       |  error404       |  ❌  | 
 error404       |  error_404      |  ❌  |  error404       |  error_404      |  ❌  |  error404       |  error_404      |  error404       |  error404       |  ✅  | 
 errCode_404    |  err_code_404   |  ❌  |  err_code_404   |  err_code_404   |  ✅  |  errCode_404    |  err_code_404   |  errCode404     |  err_code404    |  ❌  | 
 errCode404     |  err_code_404   |  ❌  |  err_code404    |  err_code_404   |  ❌  |  errCode404     |  err_code_404   |  errCode404     |  err_code404    |  ✅  | 
 err_code_404   |  errCode404     |  ✅  |  err_code_404   |  err_code404    |  ❌  |  err_code_404   |  errCode404     |  errCode404     |  err_code404    |  ❌  | 
 err_code_404   |  errCode_404    |  ❌  |  err_code_404   |  err_code_404   |  ✅  |  err_code_404   |  errCode_404    |  errCode404     |  err_code404    |  ❌  | 
 aBc            |  a_b_c          |  ✅  |  a_bc           |  a_b_c          |  ❌  |  aBc            |  a_b_c          |  aBC            |  a_bc           |  ❌  | 
 a_b_c          |  aBc            |  ✅  |  a_b_c          |  a_bc           |  ❌  |  a_b_c          |  aBc            |  aBc            |  a_bc           |  ❌  | 
 a_b_c          |  a_b_c          |  ✅  |  a_b_c          |  a_b_c          |  ✅  |  a_b_c          |  a_b_c          |  aBC            |  a_bc           |  ✅  | 
 aBC            |  a_b_c          |  ❌  |  a_bc           |  a_b_c          |  ❌  |  aBC            |  a_b_c          |  aBC            |  a_bc           |  ✅  | 
 a_b_c          |  aBC            |  ✅  |  a_b_c          |  a_bc           |  ❌  |  a_b_c          |  aBC            |  aBC            |  a_bc           |  ❌  | 


## Reproducer

I took the code from the current [normalize](https://github.com/twigphp/Twig/blob/3.x/src/Util/CallableArgumentsExtractor.php) method, the two new toSnake and toCamel from [your changes](https://github.com/twigphp/Twig/blob/3149ab346f29fa67f959e38d7628152dd33babfe/src/Util/CallableArgumentsExtractor.php) and the [snake and camel](https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/String/AbstractUnicodeString.php#L155) methods from Symfony

You can find the small script i made to test on [this gist](https://gist.github.com/smnandre/09cb10e5ce4e95fe45ce87ab7f67b14c).

**Comment by fabpot** [comment]
Your comment @smnandre made me realize that there is nothing forcing people to use proper snake or camel cases.
So, I've taken a simpler and more robust approach. I now compare the argument name and the callable parameter name after removing the `_`s and converting to lowercase. I've added all your tests to confirm that they are all behaving correctly now.

There are of course some edge cases where several parameters might resolve to the same name, but having such similar names in a signature would not be very smart :)

**Comment by smnandre** [comment]
I feared I was gonna make you lose time here .. so i'm really glad if in the end the implementation is more robust and more "tolerant" at the same time. 

🎩

File: src/Util/CallableArgumentsExtractor.php [hunk 1]
```diff
@@ -163,7 +166,17 @@ public function extractArguments(Node $arguments): array
...
+    }
+
+    // logic from Symfony\Component\String\AbstractUnicodeString::snake()
+    private function toSnakeCase(string $str): string
+    {
+        $camel = str_replace(' ', '', preg_replace_callback('/\b.(?!\p{Lu})/u', static function ($m) use (&$i) {
+            return 1 === ++$i ? ('İ' === $m[0] ? 'i̇' : mb_strtolower($m[0], 'UTF-8')) : mb_convert_case($m[0], \MB_CASE_TITLE, 'UTF-8');
+        }, preg_replace('/[^\pL0-9]++/u', ' ', $str)));
+
+        return mb_strtolower(preg_replace(['/(\p{Lu}+)(\p{Lu}\p{Ll})/u', '/([\p{Ll}0-9])(\p{Lu})/u'], '\1_\2', $camel), 'UTF-8');
```
Comment (nicolas-grekas) [inline src/Util/CallableArgumentsExtractor.php]: since normalizeName isn't unicode-aware, I suggest simplifying here:
```suggestion
    private function toSnakeCase(string $str): string
    {
        return strtolower(preg_replace(['/([A-Z]+)([A-Z][a-z])/', '/([a-z0-9])([A-Z])/'], '\1_\2', $str));
```

**Comment by stof** [comment]
what happens for named arguments passed to variadic functions ? AFAIU, the issue you referenced was about that case, not about the case where Twig is resolving named arguments for its callables.

**Comment by fabpot** [comment]
> what happens for named arguments passed to variadic functions ? AFAIU, the issue you referenced was about that case, not about the case where Twig is resolving named arguments for its callables.

Variadic is just one problem, which I "fixed" in the last commit. For variadics, I keep the attribute name as is.

**Comment by stof** [comment]
Not normalizing variadic arguments to snake case anymore will fix #3475 but would be a BC break for functions/templates that currently rely on this behavior (because the function implementation expects snake case but the template does not provide it).

**Comment by fabpot** [comment]
> Not normalizing variadic arguments to snake case anymore will fix #3475 but would be a BC break for functions/templates that currently rely on this behavior (because the function implementation expects snake case but the template does not provide it).

I know :)

**Comment by fabpot** [comment]
> > Not normalizing variadic arguments to snake case anymore will fix #3475 but would be a BC break for functions/templates that currently rely on this behavior (because the function implementation expects snake case but the template does not provide it).
> 
> I know :)

I should elaborate a bit more here. Even if it is a BC break, is it a problem in practice? The current behavior is to always convert to snake_case. So, for any app that runs today, developers must have used snake case names as well for their variadic names. And I bet developers also changed their templates to use snake_case as well to be consistent. Is it a fair assumption?

**Comment by fabpot** [comment]
Now, we trigger a deprecation when a name for a variadic arg is not snake-cased. That way, we're sure that the upgrade is smooth.

File: src/Util/CallableArgumentsExtractor.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Util/CallableArgumentsExtractor.php]: if we want to keep unicode casing, we should likely do it only when the charset is utf-?8?

**Comment by tacman** [comment]
Since twig 4 will allow BC breaks, perhaps this issue can be fixed for variadic functions then.
